temporal-sdk/src/main/java/io/temporal/common/converter/SearchAttributesPayloadConverter.java

```diff
@@ -0,0 +1,135 @@
+import io.temporal.api.common.v1.Payload;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.time.LocalDateTime;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class SearchAttributesPayloadConverter {
+
+  private static ObjectMapper mapper;
+  private static final Logger log = LoggerFactory.getLogger(SearchAttributesPayloadConverter.class);
+  private static final SearchAttributesPayloadConverter INSTANCE =
+      new SearchAttributesPayloadConverter();
+
+  private SearchAttributesPayloadConverter() {
+    mapper = new ObjectMapper();
+    mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+    mapper.registerModule(new JavaTimeModule());
+    mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+  }
+
+  public static SearchAttributesPayloadConverter getInstance() {
+    return INSTANCE;
+  }
+
+  public Optional<Payload> toData(Object obj) throws DataConverterException {
+    try {
+      byte[] serialized = mapper.writeValueAsBytes(obj);
+      String type = javaTypeToEncodedType(obj.getClass()).name();
+      return Optional.of(
+          Payload.newBuilder()
+              .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_JSON)
+              .putMetadata(EncodingKeys.METADATA_TYPE_KEY, ByteString.copyFromUtf8(type))
+              .setData(ByteString.copyFrom(serialized))
+              .build());
+    } catch (JsonProcessingException e) {
+      throw new DataConverterException(e);
+    }
+  }
+
+  public <T> T fromData(Payload payload) throws DataConverterException {
+    ByteString data = payload.getData();
+    String type = payload.getMetadataMap().get(EncodingKeys.METADATA_TYPE_KEY).toStringUtf8();
+    Type javaType = encodedTypeToJavaType(type);
+    if (data.isEmpty() || javaType == null) {
+      return null;
+    } else {
+      try {
+        @SuppressWarnings("deprecation")
+        JavaType reference = mapper.getTypeFactory().constructType(javaType, javaType.getClass());
+        return mapper.readValue(data.toByteArray(), reference);
+      } catch (IOException e) {
+        throw new DataConverterException(e);
+      }
+    }
+  }
+
+  private Type encodedTypeToJavaType(String type) {
+    SearchAttributeType attributeType;
+    try {
+      attributeType = SearchAttributeType.valueOf(type);
+    } catch (IllegalArgumentException e) {
+      return null;
+    }
+    switch (attributeType) {
+      case String:
+      case Keyword:
+        return String.class;
+      case Int:
+        return Integer.class;
+      case Double:
+        return Double.class;
+      case Bool:
+        return Boolean.class;
+      case Datetime:
+        return LocalDateTime.class;
+    }
+    throw new DataConverterException("Unsupported Search Attribute type: " + type);
+  }
+
+  private SearchAttributeType javaTypeToEncodedType(Class type) {
+    if (String.class.equals(type)) {
```

Comment:
From discussion with @alexshtin, we don't need this javaTypeToEncodedType at all.